File: klipper_docs_mcp/services/search_engine.py
"""Search Engine for Klipper Docs MCP Server.

Handles searching through documentation files with ranking and context extraction.
"""

import os
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from ..config import SearchConfig, StorageConfig
from ..exceptions import DocumentationNotAvailableError, SearchQueryEmptyError


@dataclass
class SearchResult:
    """A single search result."""

    rank: int
    path: str
    snippet: str


class SearchEngine:
    """Searches through documentation files."""

    def __init__(
        self,
        docs_dir: Path,
        search_config: Optional[SearchConfig] = None,
        storage_config: Optional[StorageConfig] = None,
    ):
        """Initialize the search engine.

        Args:
            docs_dir: Path to the documentation directory.
            search_config: Search configuration. If None, uses default.
            storage_config: Storage configuration for file extensions. If None, uses default.
        """
        self.docs_dir = docs_dir
        self.search_config = search_config or SearchConfig()
        self.storage_config = storage_config or StorageConfig()

    def _is_supported_file(self, filename: str) -> bool:
        """Check if file is supported for searching.

        Args:
            filename: Name of the file.

        Returns:
            True if file has supported extension.
        """
        return filename.endswith(self.storage_config.supported_extensions)

    def _extract_heading_matches(self, content: str, query: str) -> list[tuple[int, str]]:
        """Extract matches from Markdown headings.

        Args:
            content: File content to search.
            query: Search query.

        Returns:
            List of (position, snippet) tuples for heading matches.
        """
        matches = []
        heading_pattern = re.compile(
            r"^(#{1,6}\s.*" + re.escape(query) + ".*)$",
            re.MULTILINE | re.IGNORECASE,
        )

        heading_ranges = []
        for match in heading_pattern.finditer(content):
            start = max(0, match.start() - 50)
            end = min(len(content), match.end() + 50)
            snippet = content[start:end].strip()
            matches.append((match.start(), "heading", snippet))
            heading_ranges.append((match.start(), match.end()))

        return matches, heading_ranges

    def _extract_content_matches(
        self, content: str, query: str, heading_ranges: list[tuple[int, int]]
    ) -> list[tuple[int, str, str]]:
        """Extract matches from regular content (non-heading).

        Args:
            content: File content to search.
            query: Search query.
            heading_ranges: Ranges of heading matches to exclude.

        Returns:
            List of (position, type, snippet) tuples for content matches.
        """
        matches = []

        for match in re.finditer(re.escape(query), content, re.IGNORECASE):
            # Check if match overlaps with any heading
            is_heading = False
            for h_start, h_end in heading_ranges:
                if match.start() >= h_start and match.end() <= h_end:
                    is_heading = True
                    break

            if is_heading:
                continue

            snippet_length = self.search_config.snippet_length
            start = max(0, match.start() - snippet_length // 2)
            end = min(len(content), match.end() + snippet_length // 2)
            snippet = content[start:end].strip()
            matches.append((match.start(), "content", snippet))

        return matches

    def _determine_rank(self, filename_match: bool, matches: list) -> int:
        """Determine search result rank based on match type.

        Args:
            filename_match: Whether query matched in filename.
            matches: List of content matches.

        Returns:
            Rank value (1=highest, 3=lowest).
        """
        if filename_match:
            return 1
        elif any(m[1] == "heading" for m in matches):
            return 2
        else:
            return 3

    def _get_best_snippet(self, content: str, matches: list) -> str:
        """Get the best snippet for search results.

        Args:
            content: File content.
            matches: List of matches.

        Returns:
            Best snippet string.
        """
        if matches:
            matches.sort(key=lambda x: x[0])
            return matches[0][2]
        else:
            # Filename match - show file start
            snippet_length = self.search_config.snippet_length
            if len(content) > snippet_length:
                return content[:snippet_length] + "..."
            return content

    def search(self, query: str) -> list[SearchResult]:
        """Search documentation by query.

        Prioritizes matches in:
        1. Filenames
        2. Markdown headings (lines starting with #)
        3. Regular content

        Args:
            query: Search query string.

        Returns:
            List of SearchResult objects, ranked by relevance.

        Raises:
            DocumentationNotAvailableError: If docs directory doesn't exist.
            SearchQueryEmptyError: If query is empty.
        """
        if not self.docs_dir.exists():
            raise DocumentationNotAvailableError()

        if not query:
            raise SearchQueryEmptyError()

        query_lower = query.lower()
        results = []

        # Stream through files to avoid loading everything into memory
        for root, dirs, files in os.walk(self.docs_dir):
            for file in files:
                if not self._is_supported_file(file):
                    continue

                file_path = Path(root) / file
                rel_path = str(file_path.relative_to(self.docs_dir))

                try:
                    # Check filename match (highest priority)
                    filename_match = query_lower in rel_path.lower()

                    with open(file_path, "r", encoding="utf-8", errors="ignore") as f:
                        content = f.read()

                    # Extract heading matches
                    heading_matches, heading_ranges = self._extract_heading_matches(content, query)

                    # Extract content matches
                    content_matches = self._extract_content_matches(content, query, heading_ranges)

                    # Combine all matches
                    all_matches = heading_matches + content_matches

                    # Add result if we have matches or filename match
                    if all_matches or filename_match:
                        rank = self._determine_rank(filename_match, all_matches)
                        snippet = self._get_best_snippet(content, all_matches)
                        results.append(SearchResult(rank=rank, path=rel_path, snippet=snippet))

                except (IOError, UnicodeDecodeError):
                    continue

        # Sort by rank and limit results
        results.sort(key=lambda x: x.rank)
        max_results = self.search_config.max_results
        return results[:max_results]

    def format_results(self, results: list[SearchResult]) -> str:
        """Format search results for display.

        Args:
            results: List of SearchResult objects.

        Returns:
            Formatted string output.
        """
        if not results:
            return "No results found."

        output = []
        for result in results:
            output.append(f"## {result.path}\n{result.snippet}\n")

        return "\n".join(output)

File: src/services/SearchEngine.ts
import fs from "node:fs/promises";
import path from "node:path";
import { MAX_SEARCH_RESULTS, SNIPPET_LENGTH, SUPPORTED_EXTENSIONS } from "../config.js";
import {
  DocumentationNotAvailableError,
  SearchQueryEmptyError,
} from "../errors.js";

export interface SearchResult {
  rank: number;
  path: string;
  snippet: string;
}

export class SearchEngine {
  private docsDir: string;

  constructor(docsDir: string) {
    this.docsDir = docsDir;
  }

  private isSupportedFile(filename: string): boolean {
    return SUPPORTED_EXTENSIONS.some((ext) => filename.endsWith(ext));
  }

  private escapeRegExp(string: string): string {
    return string.replace(/[.*+?^${}()|[\\]/g, "\\$& ");
  }

  private extractHeadingMatches(
    content: string,
    query: string
  ): { matches: { start: number; snippet: string }[]; ranges: [number, number][] } {
    const matches: { start: number; snippet: string }[] = [];
    const ranges: [number, number][] = [];
    
    // JS Regex for multiline headings containing query
    // ^(#{1,6}\s.*query.*)$ with 'm' flag for multiline and 'i' for case insensitive
    const escapedQuery = this.escapeRegExp(query);
    const regex = new RegExp(`^(#{1,6}\s.*${escapedQuery}.*)$`, "gmi");

    let match;
    while ((match = regex.exec(content)) !== null) {
      const start = Math.max(0, match.index - 50);
      const end = Math.min(content.length, match.index + match[0].length + 50);
      const snippet = content.slice(start, end).trim();
      
      matches.push({ start: match.index, snippet: "heading" }); // Type marker usually not needed if we just need snippet string, but Python code used it.
                                                                // Wait, Python code returns (start, "heading", snippet)
      // Actually, I'll stick to returning what's needed for logic.
      // Python logic: matches.append((match.start(), "heading", snippet))
      
      matches.push({ start: match.index, snippet });
      ranges.push([match.index, match.index + match[0].length]);
    }

    return { matches, ranges };
  }

  private extractContentMatches(
    content: string,
    query: string,
    headingRanges: [number, number][]
  ): { start: number; snippet: string }[] {
    const matches: { start: number; snippet: string }[] = [];
    const escapedQuery = this.escapeRegExp(query);
    const regex = new RegExp(escapedQuery, "gi");

    let match;
    while ((match = regex.exec(content)) !== null) {
      // Check overlap
      let isHeading = false;
      for (const [hStart, hEnd] of headingRanges) {
        if (match.index >= hStart && match.index + match[0].length <= hEnd) {
          isHeading = true;
          break;
        }
      }

      if (isHeading) continue;

      const start = Math.max(0, match.index - SNIPPET_LENGTH / 2);
      const end = Math.min(content.length, match.index + match[0].length + SNIPPET_LENGTH / 2);
      const snippet = content.slice(start, end).trim();
      
      matches.push({ start: match.index, snippet });
    }

    return matches;
  }

  private determineRank(filenameMatch: boolean, hasHeadingMatches: boolean): number {
    if (filenameMatch) return 1;
    if (hasHeadingMatches) return 2;
    return 3;
  }

  async search(query: string): Promise<SearchResult[]> {
    if (!query) {
      throw new SearchQueryEmptyError();
    }

    try {
      await fs.access(this.docsDir);
    } catch {
      throw new DocumentationNotAvailableError();
    }

    const queryLower = query.toLowerCase();
    const results: SearchResult[] = [];

    // Recursive walk
    const walk = async (dir: string) => {
      const entries = await fs.readdir(dir, { withFileTypes: true });
      for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);
        
        if (entry.isDirectory()) {
          await walk(fullPath);
          continue;
        }

        if (!this.isSupportedFile(entry.name)) continue;

        const relPath = path.relative(this.docsDir, fullPath);
        const filenameMatch = relPath.toLowerCase().includes(queryLower);

        try {
          const content = await fs.readFile(fullPath, "utf-8");
          
          const { matches: headingMatches, ranges: headingRanges } = this.extractHeadingMatches(content, query);
          const contentMatches = this.extractContentMatches(content, query, headingRanges);

          const allMatches = [...headingMatches, ...contentMatches];
          
          if (allMatches.length > 0 || filenameMatch) {
            allMatches.sort((a, b) => a.start - b.start);
            
            const hasHeadingMatches = headingMatches.length > 0;
            const rank = this.determineRank(filenameMatch, hasHeadingMatches);
            
            // Best snippet: first match or start of file
            let snippet = "";
            if (allMatches.length > 0) {
                snippet = allMatches[0].snippet;
            } else {
                snippet = content.slice(0, SNIPPET_LENGTH) + "...";
            }

            results.push({ rank, path: relPath, snippet });
          }

        } catch (e) {
            // Ignore read errors
        }
      }
    };

    await walk(this.docsDir);

    results.sort((a, b) => a.rank - b.rank);
    return results.slice(0, MAX_SEARCH_RESULTS);
  }

  formatResults(results: SearchResult[]): string {
    if (results.length === 0) return "No results found.";
    return results.map(r => `## ${r.path}
${r.snippet}
`).join("\n");
  }
}
